### PR TITLE
HRINT-356 Node.js demos - Basics and Related Categories

### PR DESCRIPTION
**Pull request by Danielle9897**

### CODE REVIEW

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                "",
+                "* Once saveChanges() returns, the new document is persisted to the database."
+            ],
+            "Assets": [
+                {
+                    "Title": "The SaveChanges Command",
+                    "Type" : "link",
+                    "Url"  : "https://github.com/ravendb/book/blob/v4.0/Ch02/Ch02.md#savechanges"
+                }
+                //TODO
```
[the remark this answers is not in the document]
Comment (topalach) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json]: No worries, this one will work in the demo parser. This is fine.

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            "DescriptionLines": [
+                "* Create a Node.js object to be saved.",
+                "",
+                "* _The Node.js class name_ determines the collection to which the document will belong to in the database.<br>",
+                "  Collections are used to group documents together so that it is convenient to apply some operation to them.",
+                "",
+                "* _The Node.js class members_ determine the document JSON structure in the database.<br>",
+                "  The JSON serialization for the document fields follows the conventions set on the Document Store",
+                "",
+                "* Note: Having an 'Id' field property in the class definition is Not mandatory.<br>",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json]: done

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        {
+            "Title": "Create a new Node.js object to be saved",
+            "Slug" : "step-1",
+            "DescriptionLines": [
+                "* Create a Node.js object to be saved.",
+                "",
+                "* _The Node.js class name_ determines the collection to which the document will belong to in the database.<br>",
+                "  Collections are used to group documents together so that it is convenient to apply some operation to them.",
+                "",
+                "* _The Node.js class members_ determine the document JSON structure in the database.<br>",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json]: done

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    ],
+
+    "Walkthroughs": [
+        {
+            "Title": "Create a new Node.js object to be saved",
+            "Slug" : "step-1",
+            "DescriptionLines": [
+                "* Create a Node.js object to be saved.",
+                "",
+                "* _The Node.js class name_ determines the collection to which the document will belong to in the database.<br>",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json]: done

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                    "Url"  : "https://ravendb.net/docs/article-page/4.2/nodejs/studio/database/documents/documents-and-collections"
+                }
+            ]
+        },
+        {
+            "Title": "Store the new company object in the Session",
+            "Slug" : "step-2",
+            "DescriptionLines": [
+                "* Associate the new entity with the Session `store()` command.<br>",
+                "  Note: The new document will be persisted to the database only later - when saveChanges() is called.",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json]: done (but without making it bold)

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            "Title": "The new document ID is immediately available",
+            "Slug" : "step-3",
+            "DescriptionLines": [
+                "* The Session takes care of generating the document identifier automatically.<br>",
+                "  The new stored entity ID is now available and can be used.",
+                "",
+                "* _If your Node.js class was defined with an 'id' property_:<br>",
+                "  => You can get the new stored entity ID as shown in the above example.",
+                "",
+                "* _If your Node.js class was Not defined with an 'id' property_:<br>",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json]: done

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                "  => You can get the new stored entity ID using: `session.advanced().getDocumentId(newCompany)`"
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json]: done - removed ()

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/deleteDocument/metadata.json [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            "DescriptionLines": [
+                "* Mark the entity to be deleted with the Session `delete()` command.<br>",
+                "  The document is actually deleted from the database only when saveChanges() is called later on.",
+                "",
+                "_Note_:",
+                "",
+                "* When using delete() as above, with the specific ID param overload, the document is deleted with no concurrency checks.",
+                "* A _Concurrency Exception_ can be thrown only when using the following overloads:",
+                "",
+                "     * <T> void delete(T entity);",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/deleteDocument/metadata.json]: 1. Right - need to change to node.js syntax.
    I would have liked to change to:
```
  "     * delete<TEntity extends object>(entity: TEntity): Promise<void>",
  "     * delete<TEntity extends object>(id: string, expectedChangeVector: string): Promise<void>;"
```

**BUT**....I just realize that:

2. There is an issue with the parser (?) -
    whatever is in the `< >` characters will Not show in the UI.
    See attached image.   
    @topalach can you look ?

3. So for now I am just using:
```
  "     * delete(entity)",
  "     * delete(id, expectedChangeVector)"
```

[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/deleteDocument/metadata.json]: image

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/editDocument/metadata.json [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    ],
+
+    "Walkthroughs": [
+        {
+            "Title": "Load an existing document",
+            "Slug" : "step-1",
+            "DescriptionLines": [
+                "* Load an existing document from the Document Store using the Session `load()` command.<br>",
+                "",
+                "* The loaded document is added to the internal entities Map that the Session manages.<br>",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/editDocument/metadata.json]: done

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            "Url"  : "https://github.com/ravendb/book/blob/v4.0/Ch02/Ch02.md#your-first-ravendb-program"
+        }
+    ],
+
+    "Walkthroughs": [
+        {
+            "Title": "Create a Document Store Instance",
+            "Slug" : "step-1",
+            "DescriptionLines": [
+                "* The const `documentStore` will hold an instance of the _Document Store object_ that will be used across your client application.",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json]: done

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+                "  * Export & Import database data.",
+                "",
+                "* At this point, before initializing the Document Store (in step 2), the following info can be configured:",
+                "",
+                "  * _URLs_ - Specify a list of URL addresses pointing to the RavenDB server nodes composing your cluster.",
+                "    Note: Do not open a DocumentStore using URL addresses that point to nodes outside your cluster.",
+                "",
+                "  * _Database_ - Specify the default database name this Document Store will work against.",
+                "",
+                "  * _Certificate_ - Specify the Client Certificate ( *.pfx ) which is used to authenticate the client to the RavenDB server.",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json]: done - added *.pem

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                "",
+                "  * _Conventions_ - Customize various settings and behaviors."
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json]: 1. It is not an error - 'behaviors' can be plural - if it is **countable** actions
2. Online docs has :
    ...The conventions apply to many different client **behaviors**...
     https://ravendb.net/docs/article-page/4.2/Csharp/client-api/configuration/conventions
3. We could change to:
    "...**Customize various settings and operation/work modes**..."

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+                    "Type" : "link",
+                    "Url"  : "https://ravendb.net/docs/article-page/4.2/nodejs/start/getting-started"
+                }
+            ]
+        },
+        {
+            "Title": "Initialize the Document Store object",
+            "Slug" : "step-2",
+            "DescriptionLines": [
+                "* To be able to work on the DocumentStore, Call `initialize()` to get the fully initialized Document Store instance.",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json]: done

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+                    "Url"  : "https://ravendb.net/docs/article-page/4.2/nodejs/start/getting-started"
+                }
+            ]
+        },
+        {
+            "Title": "Export the Document Store Object",
+            "Slug" : "step-3",
+            "DescriptionLines": [
+                "* Export the document store object wrapped as a module.<br>",
+                "  It will be cached and act as a singleton."
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/theDocumentStore/metadata.json]: done - removed

File: DemoServer/AdditionalLanguages/nodejs/demo/relatedDocuments/createRelatedDocuments/metadata.json [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+            ]
+        },
+        {
+            "Title": "Reference the related documents",
+            "Slug" : "step-4",
+            "DescriptionLines": [
+                "* Create a reference between the primary document and the related documents.<br>",
+                "  A reference to another document is a string property that contains the ID of the related document.",
+                "",
+                "* _supplier.Id_ & _category.Id_ contain the related documents IDs that were generated by the RavenDB client in the previous step."
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/relatedDocuments/createRelatedDocuments/metadata.json]: done

File: DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+            ]
+        },
+        {
+            "Title": "The new document ID is immediately available",
+            "Slug" : "step-3",
+            "DescriptionLines": [
+                "* The Session takes care of generating the document identifier automatically.<br>",
+                "  The new stored entity ID is now available and can be used.",
+                "",
+                "* _If your Node.js class was defined with an 'id' property_:<br>",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/nodejs/demo/basics/createDocument/metadata.json]: done

**Pull request merged**